## Assignment 1 Python Basics, Built-in Structures, Functions, Files, Numpy

This assignment will contain 5 questions with details as below. The due date is September 25 (Friday), 2020 23:59PM. Each late day will result in 20% loss of total points.

### Question 1 (20 points) Datacamp courses

Datacamp is the online learning platform for data science courses and modules. As the only collaborator in Portugal, this course provides free access to the students in Business Analytics as the complementary resource. For this assignment, you are expected to complete the following moduels and submit your completion badges:

- [Introduction to Python](https://app.datacamp.com/learn/courses/intro-to-python-for-data-science)
- [Data Types for Data Science in Python](https://app.datacamp.com/learn/courses/data-types-for-data-science-in-python)
- [Python Data Science Toolbox (Part 1)](https://app.datacamp.com/learn/courses/python-data-science-toolbox-part-1)
- [Python Data Science Toolbox (Part 2)](https://app.datacamp.com/learn/courses/python-data-science-toolbox-part-2)

#### Badges
- [Introduction to Python](Badge1.pdf)
- [Data Types for Data Science in Python](Badge2.pdf)
- [Python Data Science Toolbox (Part 1)](Badge3.pdf)
- [Python Data Science Toolbox (Part 2)](Badge4.pdf)

Each badge will account for 5 points. If you are confident with the knowledge, you may skip the videos and directly finish the challenges.

### Question 2 (5 points) Geohashing

Geohashing is an outdoor sport activity played around the world since 2008. Contestents need to adventure to a geocoordinate generated by an algorithm that derives randomness from stock market data (as you should all know Eugene Fama's EMH). 

The algorithm is explained as below from 
[XKCD](https://xkcd.com/426/):
![](https://imgs.xkcd.com/comics/geohashing.png)


It described a way for a computer to create an algorithm that could generate random Global Positioning System (GPS) coordinates each day based on the Dow Jones Industrial Average and the current date.

Now your task is to decode the geohashed secret location by using Nova SBE's geocoordinate, current date (September 18, 2020) as well as the opening of DJIA on that day. Luckily, Python already has a library to calculate the geohasing using the `geohash` function in `antigravity` package:


```python
from antigravity import geohash
geohash(37.421542, -122.085589, b'2005-05-26-10458.68')
```
The first and second parameter takes the lat/lon of your preferred location and the third parameter takes the date as well as the opening of DJIA on that day, connected by `-`.

Now calculate the geohashed location using Nova SBE's geocoordinates (38.678223, -9.326266), and the current date Setember 22, 2020 and the opening of the DJIA with two decimals (you can find this at any finance news and thank you Jerome Powell!!!). 

Once you decode the location, go to google map, find out the address and write it down in the answer.

We will use the ```geohash``` function in ```antigravity``` package to calculate the geocoordinates:

In [29]:
# Question 2
from antigravity import geohash

geohash(38.678223, -9.326266, b'2020-09-22-27170.34')

38.959822 -9.819527


You can go to Google Map and type the obtained the geocoordinates and show the address of the geocoordinates

#### ANSWER
The coordinates that I obtained are **(38.959822 -9.819527)** which is in the Portuguese Seashore so I couldn't get any address

### Question 3 (15 points) Find the truth of the universe

The number $e$, known as [Euler's number](https://en.wikipedia.org/wiki/E_(mathematical_constant)) (named after the Swiss mathematician Leonhard Euler), is a mathematical constant.  It is the base of the natural logarithm. It is the limit of $(1 + \frac{1}{n})^n$ as $n$ approaches infinity, an expression that arises in the study of compound interest, as your all know from your economics/finance 101 course.

Back in 18th century, Euler derived this number through the following equation:

$e = \frac{1}{0!}+\frac{1}{1!}+\frac{1}{2!}+\frac{1}{3!}+\frac{1}{4!}\dots+\frac{1}{n!}$

Now let's stand on the giant's shoulder to get our own $e$. To do so, you need to write a function ```factorial``` to calculate the factorial number given the $n$ and then use ```for loop``` to generate the $e$. Let $n$ go from 0 to 10, and save the derived value of $e$ in a list and print it out when n is an odd number (up to 10).

Hints: 
Consider use range(#begin, #end) method

In [30]:
# Question 3
def factorial(n):
    if n == 0:
        return 1
    prod = 1
    for i in range(1, n+1):
        prod = prod * i
    return prod

euler_numbers = []
e = 0
for i in range(0,11):
    e += 1/factorial(i)
    euler_numbers.append(e)
    if(i%2 != 0):
        print(i,":", euler_numbers[i])


1 : 2.0
3 : 2.6666666666666665
5 : 2.7166666666666663
7 : 2.7182539682539684
9 : 2.7182815255731922


### Question 4 (20 points) The monologue of Macbeth

The file ```macbeth.txt``` contains the act of Macbeth, one of the greatest pieces of William Shakespeare. Read the file as we show in the lecture and answer the following question. 

1. Print the number of lines in the file
2. Identify how many times that Macebeth narrated, that is the number of times the text shows ```MACBETH``` (in upper case), not the number of times ```Macbeth```  appears in the text.
3. Slice the file with content only from the monologue of Macbeth and save it to a new file ```macbeth_monologue.txt```. 

Hints: you can check each line of the text and start to save the text to a new list when the line contains `MACBETH` and stop to save when the line contains another line of character name.

**Note** If you are using Google Colab, you need first to upload this file to the computing environment using the following code:
```python
from google.colab import files

files.upload()
```

Then it will ask you to upload the file, select the file and it will be uploaded. You can verify whether it has been successfully uploaded at the left of the notebook `Files`

In [31]:
# Question 4
#Create String with Macbeth uppercased
mac = "macbeth".upper()

lines = [mac.rstrip() for mac in open("macbeth.txt")]

narration=[]
flag = False
macbeth_count = 0
count_lines = 0

for i in lines:
    #Increase line counter but not when the line is empty
    if i != "":
        count_lines +=1
    #Check if the line only has MACBETH written
    if i == mac:
        flag = True
        macbeth_count+=1
        continue
    #Check if anyone else is narrating so it sets the flag to false meaning that macbeth stopped narration
    elif i.upper() == i:
        flag = False
    elif flag:
        narration.append(i)
    else:
        continue   

#Question 4.1
print("Number of Lines:", count_lines)
#Question 4.2
print("Macbeth narrated",macbeth_count, "times")

#Loop through the narration list to remove the empty lines
narration = [i for i in narration if i]

#Question 4.3
#Write in a new file all Macbeth's narration
with open('macbeth_monologue.txt', 'w') as f:
    for item in narration:
        f.write("%s\n" % item)

Number of Lines: 3235
Macbeth narrated 145 times


### Question 5 (40 points)  Is it Instagrammable?

Consider any photo you take at Carcavelos beach:

![carcavelos](carcavelos.jpg)

<br>

An image is composed of three matrixes, each one for an RGB (red, green and blue) channel. Each matrix has values between 0 and 255.


![image](https://static.packt-cdn.com/products/9781789613964/graphics/e91171a3-f7ea-411e-a3e1-6d3892b8e1e5.png)

On this exercise, you will apply a filter to the image in a process called convolution. This process uses the filter as a matrix that is multiplied as a sliding window to an image. On the following animation, the filter is represented as yellow, the image channel as green, and the convolution result as red.

![convolution](https://miro.medium.com/max/1070/1*Zx-ZMLKab7VOCQTxdZ1OAw.gif)


Note that the resulted matrix has a smaller shape than the original, to ensure the same shape consider a padding with value 0 that enables the filter to slide throw every pixel value as shown in the following image:

<br>

![padding](https://media5.datahacker.rs/2018/11/sl_1.png)

<br><br>

This question requires the `numpy` and `Pillow` libraries that may already be installed on your enviroment. However, if you have a problem importing this libraries, install it via pip. Example for Pillow: `pip install Pillow` 

<br>

Consider the following code to convert an image into a numpy array:

```python
from PIL import Image
from numpy import asarray

# load the image
image = Image.open('carcavelos.jpg')

# convert image to numpy array
data = asarray(image)
```



You can save a numpy array as an image with the following code:

```python
"""
color image
"""

# convert numpy array to image
img = Image.fromarray(data, 'RGB')

# save image and .png
img.save('image.png')

"""
grayscale image
"""

# convert numpy array to image
data = data.astype(np.uint8)
img = Image.fromarray(data)

# save grayscale image as .png
img.save('image.png')
```

In [32]:
from PIL import Image
from numpy import asarray
import numpy as np

# load the image
image = Image.open('carcavelos.jpg')

# convert image to numpy array
data = asarray(image)



**Question 5.1** (5 points) What is the shape of the image? Answer with the shape and indicate what each dimension represents.

In [33]:
#Shape values means (Row, Columns, Number of Matrix)
print("Rows:", data.shape[0],"\nColumns:", data.shape[1], "\nNumber of Matrix:", data.shape[2])

Rows: 816 
Columns: 1226 
Number of Matrix: 3


#### Answer
The shape function returns 3 values:
- The First One means the Number of Pixels on the Height
- The Second One means the Number of Pixels on the Width
- The Third One means the Number of Matrix we have, in this case 3 because its one for the Red Color, one for the Green and one for the Blue

**Question 5.2** (15 points) Save the "carcavelos.jpg" image as a grayscale image into "carcavelos-grayscale.png". The image should result from a numpy array with only one channel instead of 3, consider the following weights for each channel: 0.30*R + 0.59*G + 0.11*B

In [60]:
#Store on each variable the 2D Matrix correspondent to its color
r, g, b = data[:,:,0], data[:,:,1], data[:,:,2]

#Calculate the new Matrix based on the Weights given
gray = 0.2989 * r + 0.5870 * g + 0.1140 * b

#Save the image
gray = gray.astype(np.uint8)
img_grayscale = Image.fromarray(gray)
img_grayscale.save('carcavelos-grayscale.png')

**Question 5.3** (25 points) Apply convolution operations as described above to "carcavelos.jpg". 
1. Save an image into "carcavelos-sharpen.png" with the applied filter: `np.array([[0,-1,0],[-1,5,-1],[0,-1,0]])`
2. Save an image into "carcavelos-blur.png" with the applied filter: `np.array([[1,1,1],[1,1,1],[1,1,1]])/9`

In [59]:
#Function to apply the filter and save the image with the name given 
def apply_filter(filter, name):
    image = Image.open('carcavelos.jpg')
    data_image = asarray(image)

    #Store the Matrix of Red
    r = data_image[:,:,0]
    #Store the Matrix of Green
    g = data_image[:,:,1]
    #Store the Matrix of Blue
    b = data_image[:,:,2]

    #Function to apply the convolution process to a Matrix based on the filter given
    def convolution(matrix,filt):
        output = matrix*0
        Filtr,Filtc= filt.shape
        filtr = Filtr // 2 #filter center
        filtc = Filtc // 2
        for i in range(-filtr ,filtr + 1):
            matrixr = np.roll(matrix,i,axis = 0)
            for j in range(-filtc,filtc+1):
                matrixrc = np.roll(matrixr,j,axis = 1)
                output = output + filt[i + filtr ,j + filtc] * matrixrc
        return output


    #Apply the convolution to all the 3 Matrixes 
    data_image[:,:,0] = convolution(r, filter)
    data_image[:,:,1] = convolution(g, filter)
    data_image[:,:,2] = convolution(b, filter)   
    
    #Save the image
    img = Image.fromarray(data_image, 'RGB')
    nm = name + '.png'
    img.save(nm)

#Sharpen Filter
sharpen = np.array([[0,-1,0],[-1,5,-1],[0,-1,0]])
#Blur Filter
blur = np.array([[1,1,1],[1,1,1],[1,1,1]])/9

#Call the function based in the filter we want to use
apply_filter(sharpen, "carcavelos-sharpen")
apply_filter(blur, "carcavelos-blur")

